### PR TITLE
Collect module EEPROM data in dump

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
#### What I did

Add a new CLI sfputil eeprom-hexdump-all to dump all pages for module
Support collecting module EEPROM in show techsupport command

#### How I did it

HLD: https://github.com/sonic-net/SONiC/pull/1476

#### How to verify it

Manual test
Unit test
Regression test

#### New command output (if the output of a command-line utility has changed)

```
EEPROM hexdump for module 1
        Lower page 0h
        00000000 11 08 06 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000010 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000020 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000030 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000040 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000050 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000060 00 00 00 00 00 00 00 00  00 00 00 00 00 01 08 00 |................|
        00000070 00 10 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        Upper page 0h
        00000000 11 08 06 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000010 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000020 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000030 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000040 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000050 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000060 00 00 00 00 00 00 00 00  00 00 00 00 00 01 08 00 |................|
        00000070 00 10 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        Page 1h
        00000000 11 08 06 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000010 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000020 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000030 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000040 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000050 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
        00000060 00 00 00 00 00 00 00 00  00 00 00 00 00 01 08 00 |................|
        00000070 00 10 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |................|
...
EEPROM hexdump for module 2
...
```

### CODE REVIEW

**Comment by prgeor** [comment]
@Junchao-Mellanox could you please update the command ref doc https://github.com/sonic-net/sonic-utilities/blob/master/doc/Command-Reference.md

**Comment by Junchao-Mellanox** [comment]
Hi @prgeor , I checked command reference doc, it mentioned:
```
  3. sonic-clear, sfputil, etc., This document does not explain these scripts also.
```
So, it seems that we don't add doc for sfputil.

**Comment by prgeor** [comment]
> Hi @prgeor , I checked command reference doc, it mentioned:
> 
> ```
>   3. sonic-clear, sfputil, etc., This document does not explain these scripts also.
> ```
> 
> So, it seems that we don't add doc for sfputil.

@Junchao-Mellanox we can remove that statement. Recently we added sfputil firmware download command to this doc. It will be good to capture all CLIs in this Repo in one doc

File: sfputil/main.py [hunk 1]
```diff
@@ -821,6 +821,33 @@ def hexdump(indent, data, mem_address):
...
+    lines = []
+
+    for index, sfp in enumerate(platform_chassis.get_all_sfps()):
+        try:
+            presence = sfp.get_presence()
+            if not presence:
+                lines.append(f'\nModule {index + 1} not present')
+            else:
+                lines.append(f'\nEEPROM hexdump for module {index + 1}')
+                eeprom_data = sfp.dump_eeprom()
```
Comment (prgeor) [inline sfputil/main.py:838]: @Junchao-Mellanox where is dump_eeprom() defined?

File: sfputil/main.py [hunk 2]
```diff
@@ -821,6 +821,33 @@ def hexdump(indent, data, mem_address):
 
     return result
 
+
+# 'eeprom-hexdump-all' subcommand
+@show.command()
+def eeprom_hexdump_all():
```
Comment (prgeor) [inline sfputil/main.py:827]: @Junchao-Mellanox please make use of existing CLI `sfputil show eeprom-hexdump`

**Review by keboliu — APPROVED** [review]

File: sfputil/main.py [hunk 3]
```diff
@@ -729,6 +738,134 @@ def eeprom_hexdump(port, page):
...
+
+
+def dump_sfp_eeprom(sfp, page):
+    api = sfp.get_xcvr_api()
+    if api is None:
+        return f'{EEPROM_DUMP_INDENT}N/A\n'
+
+    from sonic_platform_base.sonic_xcvr.api.public import sff8636, sff8436, cmis, sff8472
+    from sonic_platform_base.sonic_xcvr.fields import consts
+    if isinstance(api, cmis.CmisApi):
```
Comment (prgeor) [inline sfputil/main.py:783]: @Junchao-Mellanox don't need to import this mid-way through the file, instead we can read the ID directly to know the API as done in existing code to choose between eeprom_hexdump_sff8472() and eeprom_hexdump_sff8636()

**Review by prgeor — COMMENTED** [review]
@Junchao-Mellanox I would suggest to enhance existing APIs: eeprom_hexdump_sff8472() to eeprom_hexdump_sff8472(port, physical_port, page`s`). Let the caller take care of what pages or list of pages to pass based upon if the cable support only flat memory, if cable is CMIS, SFF etc.. This way we can reuse existing API. You can take approach of is_xcvr_optical() (see sonic-mgmt repo) to know if SFP+ is copper or not.

**Comment by Junchao-Mellanox** [comment]
> @Junchao-Mellanox I would suggest to enhance existing APIs: eeprom_hexdump_sff8472() to eeprom_hexdump_sff8472(port, physical_port, page`s`). Let the caller take care of what pages or list of pages to pass based upon if the cable support only flat memory, if cable is CMIS, SFF etc.. This way we can reuse existing API. You can take approach of is_xcvr_optical() (see sonic-mgmt repo) to know if SFP+ is copper or not.

Hi @prgeor ,
I was thinking of reusing eeprom_hexdump_sff8472. But I found it is very difficult. eeprom_hexdump_sff8472 is designed for single port:
1. There are many code like `sys.exit(ERROR_NOT_IMPLEMENTED)`. It would exit the CLI if any port got unexpected result. This is not expected for multi ports dump.
2. It dumps EEPROM for logical port like Ethernet0, Ethernet4. This is OK for single port, but not so good for multi port dump. For multi port dump, I choose to dump eeprom for physical port so that split port won't be dump many times.

I understand your point is to reuse existing code and keep the design clean. I totally agree with that. So, how about this:
I will totally replace eeprom_hexdump_sff8472 with dump_eeprom_pages_sff8472 in a future PR. I don't want to do it in current PR because I don't want to take the risk to break existing function.

File: sfputil/main.py (hunk 3 above)
Comment (Junchao-Mellanox) [reply sfputil/main.py:783]: Hi @prgeor , I think it is better way to do it like this. For example, say we have a new CMIS cable type whose ID is 0x34(just example). `if isinstance(api, cmis.CmisApi):` does not require any change; but reading ID solution will require a PR to change the sonic utilities to take the new ID 0x34 into account.

File: sfputil/main.py [hunk 4]
```diff
@@ -729,6 +738,134 @@ def eeprom_hexdump(port, page):
...
+            if api.is_flat_memory():
+                pages = [0]
+            else:
+                pages = [0, 1, 2, 3]
+        else:
+            pages = [page]
+        return dump_eeprom_pages(sfp, pages)
+    elif isinstance(api, sff8472.Sff8472Api):
+        if page is None:
+            is_active_cable = api.xcvr_eeprom.read(consts.SFP_CABLE_TECH_FIELD) == 'Active Cable'
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply sfputil/main.py:808]: I will fix this

File: sfputil/main.py (hunk 3 above)
Comment (Junchao-Mellanox) [reply sfputil/main.py:783]: I will double confirm that isinstance works for ccmis and other

File: sfputil/main.py [hunk 5]
```diff
@@ -678,148 +681,281 @@ def eeprom(port, dump_dom, namespace):
...
+        overall_offset: overall offset in flat memory
+        size: size of bytes to be dumped
+        page_offset: offset within a page, only for print purpose
+        no_format: False if dump with hex format else dump with flat hex string. Default False.
+
+    Returns:
+        tuple(0, dump string) if success else tuple(error_code, error_message)
+    """
+    sfp = platform_chassis.get_sfp(physical_port)
+    page_dump = sfp.read_eeprom(overall_offset, size)
```
Comment (prgeor) [inline sfputil/main.py:914]: @Junchao-Mellanox just a suggestion, if you could rename overall_offset -> flat_offset that will reflect this offset is considered when viewing the EEPROM a single contigious FLAT memeory

File: sfputil/main.py [hunk 6]
```diff
@@ -678,148 +681,281 @@ def eeprom(port, dump_dom, namespace):
...
+                pages = [0]
+                if page not in pages:
+                    pages.append(page)
+            return eeprom_hexdump_pages_general(logical_port_name, pages, page)
+        elif isinstance(api, sff8636.Sff8636Api) or isinstance(api, sff8436.Sff8436Api):
+            if page is None:
+                if api.is_flat_memory():
+                    pages = [0]
+                else:
+                    pages = [0, 1, 2, 3]
```
Comment (prgeor) [inline sfputil/main.py:789]: @Junchao-Mellanox can we define this list of pages as SFF8436_MODULE_PAGE = [] at the top?

File: sfputil/main.py [hunk 7]
```diff
@@ -830,6 +960,7 @@ def convert_byte_to_valid_ascii_char(byte):
     else:
         return chr(byte)
 
+
```
Comment (mihirpat1) [inline sfputil/main.py:963]: Please revert this and other unintended change.

**Review by prgeor — DISMISSED** [review]

**Comment by Junchao-Mellanox** [comment]
Hi @mihirpat1 @prgeor , could you please sign-off?

**Review by mihirpat1 — APPROVED** [review]

File: sfputil/main.py [hunk 8]
```diff
@@ -687,123 +693,247 @@ def eeprom(port, dump_dom, namespace):
...
+        click.echo('\n'.join(lines))
 
-    if platform_sfputil.is_logical_port(port) == 0:
-        click.echo("Error: invalid port {}".format(port))
-        print_all_valid_port_values()
-        sys.exit(ERROR_INVALID_PORT)
 
-    if page is None:
-        page = '0'
+def validate_eeprom_page(page):
```
Comment (prgeor) [inline sfputil/main.py:726]: @Junchao-Mellanox can we use click option to validate page?

File: doc/Command-Reference.md [hunk 9]
```diff
@@ -12914,8 +12915,89 @@ Clear MACsec counters which is to reset all MACsec counters to ZERO.
...
+  Display EEPROM hexdump of SFP transceiver(s)
+Options:
+  -p, --port <port_name>    Display SFP EEPROM hexdump for port <port_name>
+  -n, --page <page_number>  Display SFP EEEPROM hexdump for
+                            <page_number_in_hex>
+  --help                    Show this message and exit.
+```
+
+```
+admin@sonic:~$ sfputil show eeprom-hexdump --port Ethernet0 --page 0
```
Comment (prgeor) [inline doc/Command-Reference.md:12936]: @Junchao-Mellanox could you add example of upper page > 0?

File: sfputil/main.py [hunk 10]
```diff
@@ -687,123 +693,247 @@ def eeprom(port, dump_dom, namespace):
...
+                    pages = copy.deepcopy(CMIS_MODULE_PAGES)
+                    if api.is_coherent_module():
+                        pages.extend(CMIS_COHERENT_MODULE_PAGES)
+                    cdb_support = api.xcvr_eeprom.read(consts.CDB_SUPPORT)
+                    if cdb_support != 0:
+                        pages.append(0x9f)
+            else:
+                pages = [0]
+                if page not in pages:
+                    pages.append(page)
```
Comment (prgeor) [inline sfputil/main.py:794]: @Junchao-Mellanox why display upper-page 0 always even if user wants to see upper-page >] 1?

File: sfputil/main.py [hunk 11]
```diff
@@ -687,123 +693,247 @@ def eeprom(port, dump_dom, namespace):
...
+            return_code, output = eeprom_dump_general(physical_port, page, 0, PAGE_SIZE, 0)
+            if return_code != 0:
+                return return_code, output
+            lines.append(output)
+
+            lines.append(f'\n{EEPROM_DUMP_INDENT}Upper page 0h')
+            return_code, output = eeprom_dump_general(physical_port, page, PAGE_OFFSET, PAGE_SIZE, PAGE_OFFSET)
+            if return_code != 0:
+                return return_code, output
+            lines.append(output)
```
Comment (prgeor) [inline sfputil/main.py:853]: @Junchao-Mellanox can we avoid displaying upper page 0 if user input target page >= 1

File: sfputil/main.py (hunk 8 above)
Comment (Junchao-Mellanox) [reply sfputil/main.py:726]: This is for backward compatible. Before the PR, we allowed None here and translate None to 0.

File: sfputil/main.py (hunk 10 above)
Comment (Junchao-Mellanox) [reply sfputil/main.py:794]: For backward compatible, previous implementation always print page 0. As you can see, I did not change the previous unit test cases, and all passed.

File: sfputil/main.py (hunk 11 above)
Comment (Junchao-Mellanox) [reply sfputil/main.py:853]: For backward compatible, previous implementation always print page 0. As you can see, I did not change the previous unit test cases, and all passed.

File: sfputil/main.py (hunk 8 above)
Comment (prgeor) [reply sfputil/main.py:726]: @Junchao-Mellanox can't we use https://github.com/sonic-net/sonic-utilities/blob/master/sfputil/main.py#L1660 here? What backward compatibility is broken?

Comment (Junchao-Mellanox) [reply sfputil/main.py:726]: Please check this line: https://github.com/sonic-net/sonic-utilities/blob/19ea8493388536921ad204833157dac2cd2bf3ba/sfputil/main.py#L693. Page option is an optional option, so it allows None.

**Comment by keboliu** [comment]
@prgeor @mihirpat1 could you please sign-off?

**Review by mihirpat1 — APPROVED** [review]

File: sfputil/main.py (hunk 8 above)
Comment (prgeor) [reply sfputil/main.py:726]: @Junchao-Mellanox use something like below:-
`@click.option('-n', default=0, '--page', metavar='<page_number>', help="Display SFP EEEPROM hexdump for <page_number_in_hex>")`

Comment (prgeor) [reply sfputil/main.py:726]: @Junchao-Mellanox then you don't need to validate page#

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@Junchao-Mellanox please check cherry-pick conflict

**Pull request merged**